pkg/envoy/lds/egress.go

```diff
@@ -0,0 +1,124 @@
+package lds
+
+import (
+	"fmt"
+	"net"
+
+	xds_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	xds_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	xds_tcp_proxy "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/envoy/route"
+	"github.com/openservicemesh/osm/pkg/trafficpolicy"
+)
+
+const (
+	egressHTTPFilterChainPrefix = "egress-http"
+	egressTCPFilterChainPrefix  = "egress-tcp"
+)
+
+var (
+	// httpProtocols is the list of allowed HTTP protocols that the downstream can use in
+	// an HTTP request that will be subjected to HTTP routing rules.
+	// *Note: HTTP2 over TLS (h2) traffic is not subjected to HTTP based routing rules because the
+	// traffic is encrypted and will be proxied as a TCP stream instead.
+	httpProtocols = []string{"http/1.0", "http/1.1", "h2c"}
+)
+
+// getEgressFilterChainsForMatches returns a slice of egress filter chains for the given traffic matches
+func (lb *listenerBuilder) getEgressFilterChainsForMatches(matches []*trafficpolicy.TrafficMatch) []*xds_listener.FilterChain {
+	var filterChains []*xds_listener.FilterChain
+
+	for _, match := range matches {
+		switch match.DestinationProtocol {
+		case constants.ProtocolHTTP:
+			// HTTP protocol --> HTTPConnectionManager filter
+			if filterChain, err := lb.getEgressHTTPFilterChain(match.DestinationPort); err != nil {
+				log.Error().Err(err).Msgf("Error building egress HTTP filter chain for port [%d]", match.DestinationPort)
+			} else {
+				filterChains = append(filterChains, filterChain)
+			}
+
+		case constants.ProtocolTCP, constants.ProtocolHTTPS:
+			// TCP or HTTPS protocol --> TCPProxy filter
+			if filterChain, err := lb.getEgressTCPFilterChain(*match); err != nil {
+				log.Error().Err(err).Msgf("Error building egress filter chain for match [%v]", *match)
+			} else {
+				filterChains = append(filterChains, filterChain)
+			}
+		}
+	}
+
+	return filterChains
+}
+
+func (lb *listenerBuilder) getEgressHTTPFilterChain(destinationPort int) (*xds_listener.FilterChain, error) {
+	filter, err := lb.getOutboundHTTPFilter(route.GetEgressRouteConfigNameForPort(destinationPort))
```

Comment:
nit: assuming we strictly refer to outbound for in-mesh connectivity, "outbound" word use could be misleading in an "egress" function

Reply:
Will `s/getOutboundHTTPFilterChainForService/getOutboundHTTPFilterChainForMeshService/g`, that way `outbound == mesh outbound || egress`